File: site/utils/event-types.ts

export type CardID = number

export type ServerEvent = {
   event_code: ServerEventCodes,
   session_id?: string,
   client_id?: string,
   session_client_ids?: Array<string>,
}

export enum ServerEventCodes {
  ClientJoined = 1,
  ClientLeft,
  GameStarted,
  TurnStart,
  DataResponse,
}

export type ClientEvent = {
   event_code: ClientEventCodes,
   target_ids?: Array<string>,
   card_id?: CardID,
   session_id?: string,
}

export enum ClientEventCodes {
  JoinSession = 1,
  CreateSession,
  LeaveSession,
  StartGame,
  EndTurn,
  PlayCard,
  DataRequest,
}

File: server/src/event_types.rs
use serde::{Deserialize, Serialize};
use serde_repr::{Deserialize_repr, Serialize_repr};

pub type CardID = u8;

#[derive(Serialize)]
pub struct ServerEvent {
  pub event_code: ServerEventCodes,
  pub session_id: Option<String>,
  pub client_id: Option<String>,
  pub session_client_ids: Option<Vec<String>>,
}

#[derive(Serialize_repr)]
#[repr(u8)]
pub enum ServerEventCodes {
  // session_id, client_id, session_client_ids
  ClientJoined = 1,
  // client_id
  ClientLeft,
  GameStarted,
  // session_id, session_client_ids
  DataResponse,
  // session_id, client_id
  InvalidSessionID,
}

#[derive(Deserialize)]
pub struct ClientEvent {
  pub event_code: ClientEventCodes,
  pub target_ids: Option<Vec<String>>,
  pub card_id: Option<CardID>,
  pub session_id: Option<String>,
}

#[derive(Deserialize_repr)]
#[repr(u8)]
pub enum ClientEventCodes {
  // session_id
  JoinSession = 1,
  CreateSession,
  LeaveSession,
  StartGame,
  EndTurn,
  PlayCard,
  DataRequest,
}
